internal/dependency/client_set.go

```diff
@@ -120,6 +120,20 @@ func (c *ClientSet) CreateConsulClient(i *CreateClientInput) error {
 	}
 	c.Unlock()
 
+	// spin until Consul cluster has a leader
+	retryTime := time.Second
+	for {
+		leader, err := client.Status().Leader()
```

Comment:
does `Leader()` return an error if there is not a leader set? I'm cautious about waiting a minute if there were bad credentials or wrong address.

Reply:
Docs have it with 1 proper response where any error would be an HTTP error. I'm not 100% on issues related to tokens and such, though it doesn't have any special ACL requirement. I added the max time check to limit it a bit as retrying forever, as ESM does, didn't seem best.

We could lower the retry if you think a shorter time would be better. The 1 minute max test results in 5 retries, totaling about a minute of waiting.

Reply:
If it errors because Consul is unreachable, or some other related HTTP errors i think it should exit early. If it is reachable but without a Consul leader then waiting a minute seems reasonable. Though what i'm not certain of is if it is distinguishable.

Reply:
I guess I was being a little lazy and didn't want to test all the possible consul errors :pensive: ...  but I can get the basics in there where we can add more as it comes up. So I'll add checks for consul being unreachable and see about other obvious HTTP errors (probably just 500s for now).

Reply:
Another reason I probably didn't is because the create API doesn't have an error return, so there's no place to bubble up that error. I'm going to look into alternative places to add it first, otherwise I'll need to update the API.

Reply:
It is really ugly how this impacts code as it propagates up the stack. I'm going to spend some time looking into alternatives. A post run Init() could do it, but I want to look around for other patterns to see if there is something better.

Reply:
Interesting. if the API doesn't return an error when there isn't a leader address, what does it return? I tried simulating this scenario locally and it gave me `"127.0.0.1:8300"` though i'm not sure where 8300 came from when the 1 server is on 8500.

Reply:
"127.0.0.1:8300" .. That is the RPC address vs. 8500 which is the HTTP API address. I saw the same thing and just assume it is the RPC address that is getting returned there. Can't say as to why.

Reply:
https://www.consul.io/docs/install/ports.html

Reply:
> [..] if the API doesn't return an error when there isn't a leader address, what does it return?

It might return an error now? That was added in 2017 with nothing indicating why it checks that way.

Reply:
Ah thanks John for the context! This is an interesting bit of logs for `dev` mode... Looks like leader is `""` when no leader has been elected ([Learn logs](https://learn.hashicorp.com/tutorials/consul/get-started-agent#start-the-agent)) -- which you already found from checking out your new changes. Ya Kim needs to RTFLogs 😂

```
==> Log data will now stream in as it occurs:

    ...
    2019/07/15 19:24:34 [INFO]  raft: Initial configuration (index=1): [{Suffrage:Voter ID:019063f6-9215-6f2c-c930-9e84600029da Address:127.0.0.1:8300}]
    2019/07/15 19:24:34 [INFO]  raft: Node at 127.0.0.1:8300 [Follower] entering Follower state (Leader: "")
    ...
    2019/07/15 19:24:34 [WARN]  raft: Heartbeat timeout from "" reached, starting election
    2019/07/15 19:24:34 [INFO]  raft: Node at 127.0.0.1:8300 [Candidate] entering Candidate state in term 2
    ...
    2019/07/15 19:24:34 [INFO]  raft: Election won. Tally: 1
    2019/07/15 19:24:34 [INFO]  raft: Node at 127.0.0.1:8300 [Leader] entering Leader state
    2019/07/15 19:24:34 [INFO] consul: cluster leadership acquired
```